src/tinuous/util.py

```diff
@@ -46,17 +47,78 @@ def iterfiles(dirpath: Path) -> Iterator[Path]:
                 yield p
 
 
+class LazySlicingFormatter(Formatter):
+    def __init__(self, var_defs: Dict[str, str]):
+        self.var_defs: Dict[str, str] = var_defs
+        self.expanded_vars: Dict[str, str] = {}
+        super().__init__()
+
+    def get_value(
+        self, key: Union[int, str], args: Sequence[Any], kwargs: Mapping[str, Any]
+    ) -> Any:
+        if isinstance(key, int):
+            return args[key]
+        elif key in kwargs:
+            return kwargs[key]
+        elif key in self.expanded_vars:
+            return self.expanded_vars[key]
+        elif key in self.var_defs:
+            self.expanded_vars[key] = self.format(self.var_defs[key], **kwargs)
+            return self.expanded_vars[key]
+        else:
+            raise KeyError(key)
+
+    def get_field(
+        self, field_name: str, args: Sequence[Any], kwargs: Mapping[str, Any]
+    ) -> Any:
+        m = re.match(r"\w+", field_name)
+        assert m, f"format field name {field_name!r} does not start with arg_name"
+        key = m.group()
+        obj = self.get_value(key, args, kwargs)
+        s = field_name[m.end() :]
+        while s:
+            m = re.match(r"\.(?P<attr>\w+)|\[(?P<index>[^]]+)\]", s)
+            assert m, f"format field name {field_name!r} has invalid attr/index"
+            s = s[m.end() :]
+            attr, index = m.group("attr", "index")
+            if attr is not None:
+                obj = getattr(obj, attr)
+            else:
+                assert index is not None  # type: ignore[unreachable]
+                try:
+                    sl = parse_slice(index)
```

Comment:
I decided not to push on subclassing `string.Formatter` whenever we were talking about enhancing the extra formatters (like for shortened `commit`) since iirc it was lacking slicing, so here you are (re)implementing it?
Please add tests to cover `except` clause here and common cases like `[-3:]` and `[start:end]`.

On a positive side -- this would open the possibility for custom formatter directives (in addition to `r`) ;-) ! (I am still wondering about `git describe` for commits ;))